Tools/scripts/run_coverage.py

```diff
@@ -0,0 +1,275 @@
+                log_file.write(result.stdout)
+        except subprocess.CalledProcessError as err:
+            print("ERROR :")
+            print(err.cmd)
+            print(err.output)
+            exit(1)
+        self.progress("Initialization done !")
+
+    def check_build(self, name, path):
+        """Check that build directory is not empty and that binaries are built with the coverage flags."""
+        self.progress("Checking that %s are set up and built" % name)
+        if os.path.exists(path):
+            if not os.listdir(path):
+                return False
+        else:
+            return False
+        self.progress("Checking %s build configuration for coverage" % name)
+        with open(path + "/../compile_commands.json", "r") as searchfile:
+            for line in searchfile:
+                if "-ftest-coverage" in line:
+                    return True
+                else:
+                    continue
+            print("%s weren't build with coverage support" % name)
+            return False
+
+    def run_build(self):
+        """Clean the build directory and build binaries for coverage."""
+        os.environ["CCFLAGS"] = os.environ.get("CCFLAGS", "") + " -fprofile-arcs -ftest-coverage"
+        os.environ["CXXFLAGS"] = os.environ.get("CXXFLAGS", "") + " -fprofile-arcs -ftest-coverage"
+        os.environ["LINKFLAGS"] = os.environ.get("LINKFLAGS", "") + " -lgcov -coverage"
+        os.environ["COVERAGE"] = "1"
+        self.progress("Removing previous build binaries")
+        try:
+            shutil.rmtree(os.path.join(root_dir, "build"))
+        except OSError as e:
+            print("Error: %s : %s" % (self.REPORT_DIR, e.strerror))
+
+        self.progress("Building examples and SITL binaries")
+        os.chdir(root_dir)
+        waf_light = os.path.join(root_dir, "modules/waf/waf-light")
+        try:
+            subprocess.run([waf_light, "configure", "--board=linux", "--debug"], check=True)
+            subprocess.run([waf_light, "examples"], check=True)
+            subprocess.run([waf_light, "configure", "--debug"], check=True)
+            subprocess.run([waf_light], check=True)
+        except subprocess.CalledProcessError as err:
+            print("ERROR :")
+            print(err.cmd)
+            print(err.output)
+            exit(1)
+        self.progress("Build examples and vehicle binaries done !")
+
+    def run_full(self):
+        """Run full coverage on maximum of ArduPilot binaries and test functions."""
+        self.progress("Running full test suite...")
+        self.run_build()
+        self.init_coverage()
+        self.progress("Running tests")
+        SPEEDUP = 5
+        TIMEOUT = 14400
+        autotest = os.path.join(root_dir, "Tools/autotest/autotest.py")
+
+        self.progress("Running run.examples")
+        subprocess.run([autotest, "--timeout=" + str(TIMEOUT), "--debug", "--no-clean", "--speedup=" + str(SPEEDUP),
+                        "run.examples"])
```

Comment:
Please break all of these over multiple lines, as I showed in an earlier example.